### PR TITLE
SNOW-68770 .NET External Browser auth support

### PR DESCRIPTION
**Pull request by ChTimTsubasa**
Support External browser authentication

### CODE REVIEW

File: Snowflake.Data.Tests/SFConnectionIT.cs [hunk 1]
```diff
@@ -299,16 +299,36 @@ public void TestUnknownAuthenticator()
...
+        [Ignore("This test requires manual interaction and therefore cannot be run in CI")]
+        public void TestSSOConnection()
         {
             using (IDbConnection conn = new SnowflakeDbConnection())
             {
-                conn.ConnectionString = "scheme=http;host=testaccount.reg.snowflakecomputing.com;port=8082;user=qa@snowflakecomputing.com;password=Test123!;" +
-                    "account=testaccount;role=sysadmin;db=testdb;schema=public;warehouse=regress;authenticator=https://snowflakecomputing.okta.com";
+                conn.ConnectionString
+                    = ConnectionStringWithoutAuth
+                    + ";authenticator=externalbrowser;";
```
Comment (ChTimTsubasa) [inline Snowflake.Data.Tests/SFConnectionIT.cs]: I realized that you don't even need to have a user specified to have externalbrowser authentication. What are you guys' thought? My feeling as a user is that this is a little cumbersome given that I should type in my user name in the browser and also have user specified in the connection string;

File: Snowflake.Data/Core/Authenticator/ExternalBrowserAuthenticator.cs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+            // .net standard would pass here
+            Process.Start(url);
+#else
+            // hack because of this: https://github.com/dotnet/corefx/issues/10361
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                url = url.Replace("&", "^&");
+                Process.Start(new ProcessStartInfo("cmd", $"/c start {url}") { CreateNoWindow = true });
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
```
Comment (ChTimTsubasa) [inline Snowflake.Data/Core/Authenticator/ExternalBrowserAuthenticator.cs]: @howryu  Do we support linux and Mac?

File: README.md [hunk 3]
```diff
@@ -107,13 +107,13 @@ i.e "\<key1\>=\<value1\>;\<key2\>=\<value2\>...". The following table lists all
...
 | DB                  | No       |                                                                               |
 | HOST                | No       | If no value specified, driver will use \<ACCOUNT\>.snowflakecomputing.com. However, if you are not in us-west deployment, HOST is required, i.e. XXX.us-east-1.snowflakecomputing.com |
-| PASSWORD            | Yes      |                                                                               |
+| PASSWORD            | Depends  | Not required for external browser, required for other authentication methods. |
 | ROLE                | No       |                                                                               |
 | SCHEMA              | No       |                                                                               |
-| USER                | Yes      |                                                                               |
+| USER                | Depends  | Not required for external browser, but will check if presented; required for other authetication methods.|
```
Comment (smtakeda) [inline README.md]: USER is required for externalbrowser mode.

File: Snowflake.Data.Tests/SFConnectionIT.cs (hunk 1 above)
Comment (smtakeda) [reply Snowflake.Data.Tests/SFConnectionIT.cs]: User is required to match the identity. In particular, token id cache will require to match.

File: Snowflake.Data/Core/RestRequest.cs [hunk 4]
```diff
@@ -131,21 +131,44 @@ HttpRequestMessage IRestRequest.ToRequestMessage(HttpMethod method)
...
+
+    class AuthenticatorRequestData
+    {
+        [JsonProperty(PropertyName = "ACCOUNT_NAME", NullValueHandling = NullValueHandling.Ignore)]
+        internal String AccountName { get; set; }
+
+        [JsonProperty(PropertyName = "AUTHENTICATOR")]
+        internal String Authenticator { get; set; }
+
+        [JsonProperty(PropertyName = "BROWSER_MODE_REDIRECT_PORT", NullValueHandling = NullValueHandling.Ignore)]
```
Comment (smtakeda) [inline Snowflake.Data/Core/RestRequest.cs]: Got this.

File: Snowflake.Data/Core/RestResponse.cs [hunk 5]
```diff
@@ -27,60 +37,51 @@ class NullDataResponse : BaseRestResponse
...
-        [JsonProperty(PropertyName = "firstLogin", NullValueHandling = NullValueHandling.Ignore)]
-        internal bool firstLogin { get; set; }
-
-        [JsonProperty(PropertyName = "healthCheckInterval", NullValueHandling = NullValueHandling.Ignore)]
-        internal Int16 healthCheckInterval { get; set; }
-
-        [JsonProperty(PropertyName = "newClientForUpgrade", NullValueHandling = NullValueHandling.Ignore)]
-        internal string newClientForUpgrade { get; set; }
-
-        [JsonProperty(PropertyName = "sessionId", NullValueHandling = NullValueHandling.Ignore)]
```
Comment (smtakeda) [inline Snowflake.Data/Core/RestResponse.cs]: Basically I agree to remove unused parameters for now. Later when we need it, we can put them back.

File: Snowflake.Data.Tests/SFConnectionIT.cs (hunk 1 above)
Comment (smtakeda) [reply Snowflake.Data.Tests/SFConnectionIT.cs]: See this: https://github.com/snowflakedb/snowflake-connector-python/blob/master/connection.py#L756
oAuth won't take USER.

Comment (smtakeda) [reply Snowflake.Data.Tests/SFConnectionIT.cs]: More accurately LOGIN_NAME is required. So LOGIN_NAME is not necessary the same as SSO user name.
NOTE: Snowflake drivers take LOGIN_NAME as `USER`.

File: Snowflake.Data/Core/Authenticator/ExternalBrowserAuthenticator.cs (hunk 2 above)
Comment (ChTimTsubasa) [reply Snowflake.Data/Core/Authenticator/ExternalBrowserAuthenticator.cs]: Confirmed with @howryu we only claim supporting Windows.

File: Snowflake.Data.Tests/SFConnectionIT.cs (hunk 1 above)
Comment (ChTimTsubasa) [reply Snowflake.Data.Tests/SFConnectionIT.cs]: Addressed

**Pull request merged**